Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -94,6 +96,24 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
     return m_observer.get();
 }
 
+static bool targetIsSelected(const Element& target)
+{
+    VisibleSelection selection = target.document().selection().selection();
+    if (auto range = selection.range()) {
+        for (Ref node : intersectingNodes(*range)) {
+            Element* ancestor = node->parentElementInComposedTree();
+            while (ancestor) {
+                if (ancestor == &target && ancestor->renderer()) {
+                    ASSERT(ancestor->renderer()->style().contentVisibility() == ContentVisibility::Auto);
+                    return true;
+                }
+                ancestor = ancestor->parentElementInComposedTree();
+            }
+        }
+    }
+    return false;
+}
```

Comment:
Does something like this work? (walking down from the target rather than up from the intersecting node)

Maybe @rniwa / @whsieh / @megangardner have better suggestions though (we want to figure out if a given element has a selection inside)

```suggestion
static bool targetContainsSelection(const Element& target)
{
    auto selection = target.document().selection().selection();
    if (auto range = selection.range()) {
        for (Ref node : intersectingNodes(*range)) {
            if (target.containsIncludingShadowDOM(node))
                return true;
        }
    }
    return false;
}
```

Reply:
@whsieh suggested using treeOrder() for more efficiency, here's something that probably works:

```
static bool targetContainsSelection(const Element& target)
{
    auto selectionRange = target.document().selection().selection().range();
    if (!selectionRange)
        return false;
    return contains<ComposedTree>(selectionRange, makeBoundaryPointBeforeNodeContents(target)) || contains<ComposedTree>(selectionRange, makeBoundaryPointAfterNodeContents(target));
}
```

Reply:
Thanks guys! Will try this tomorrow (also looks like I need to rebase).

Reply:
The snippet with contains fails on the selection tests.
I found this works:
return intersects<ComposedTree>(*selectionRange, target);
But I am not sure if that is more efficient than using intersectionNodes.

Reply:
Let's use `return intersects<ComposedTree>(*selectionRange, target);`

It's less verbose and uses `treeOrder` as Wenson suggested.

Reply:
Sure, done.